src/VASSAL/build/module/Chatter.java

```diff
@@ -115,55 +132,214 @@ public void actionPerformed(ActionEvent e) {
+    String file, tag, replace;
+    int base;
+    while (s.toLowerCase().contains(keystring)) { // Find next key (to-lower so we're not case sensitive)
+      base = s.toLowerCase().indexOf(keystring);
+      file = s.substring(base + keystring.length(), s.length()).split("\"")[0]; // Pull the filename out from between the quotes
+      tag  = s.substring(base, base + keystring.length()) + file + "\"";    // Reconstruct the part of the tag we want to remove, leaving all attributes after the filename alone, and properly matching the upper/lowercase of the keystring						
+
+      try {
+        url = GameModule.getGameModule().getDataArchive().getURL("images/" + file);
+        replace = "<img  src=\"" + url.toString() + "\""; // Fully qualified URL if we are successful. The extra
+                                                          // space between IMG and SRC in the processed
+                                                          // version ensures we don't re-find THIS tag as we iterate
+      } catch (IOException ex) {
+        replace = "<img  src=\"" + file + "\""; // Or just leave in except alter just enough that we won't find this tag again
+      }
+
+      if (s.contains(tag)) {
+        s = s.replaceFirst(tag, replace); // Swap in our new URL-laden tag for the old one.
+      } else {
+        break; // BR// If something went wrong in matching up the tag, don't loop forever
+      }
+    }
+
+    //BR// Now we have to fix up any legacy angle brackets around the word <observer>
+    keystring = Resources.getString("PlayerRoster.observer");
+    replace = keystring.replace("<","&lt;").replace(">","&gt;");
+    if (replace != keystring) {
+      s = s.replace(keystring, replace);
+    }
+
+    //BR// Insert a div of the correct style for our line of text. Module designer
+    //BR//still free to insert <span> tags and <img> tags and the like in Report
+    //BR//messages.
+    try {
+      kit.insertHTML(doc, doc.getLength(), "\n<div class=" + style + ">" + s + "</div>", 0, 0, null);
+    } catch (BadLocationException ble) {
+      ErrorDialog.bug(ble);
+    } catch (IOException ex) {
+      ErrorDialog.bug(ex);
+    }
+    conversation.update(conversation.getGraphics()); //BR// Force graphics to update
   }
 
   /** @deprecated use GlobalOptions.getPlayerId() */
-  @Deprecated public void setHandle(String s) {
+  @Deprecated
+  public void setHandle(String s) {
   }
 
   /** @deprecated use GlobalOptions.getPlayerId() */
-  @Deprecated public String getHandle() {
+  @Deprecated
+  public String getHandle() {
     return GlobalOptions.getInstance().getPlayerId();
   }
 
+  //BR// Adds or updates a CSS stylesheet entry. Styles in the color, font type, and font size.
+  private void addStyle(String s, Font f, Color c, String font_weight, int size) {
+    if ((style == null) || (c == null)) return;
+    style.addRule(s + " {color:" + String.format("#%02x%02x%02x", c.getRed(), c.getGreen(), c.getBlue())
```

Comment:
Looks a bit wild, some strategically placed newlines would make this easier readable, e.g.
```
s
+ " color:" + x
+ "; font-family:" + y
+ "; font-size:" + z
(etc)
```

Reply:
fixed for next push